design/user-profile.md

```diff
@@ -86,27 +86,63 @@ A very rough idea on how the user profile JSON could look like:
     {
         "attributes": [{
             "name": "department",
-            "view": ["admin", "user"],
-            "edit": ["admin"],
+            "permissions": {
+                    "view": ["admin", "user"], 
+                    "edit": ["admin"],
+            }
             "requirement":"always",
             "validation": {
                 "name", 
```

Comment:
@stianst we experience some issues regarding the topic of validation.
While browsing the codebase it turns out there are several completely incompatible ideas of validation right now:
UpdateProfile --> The profile is not validated as in the registration but some assessments are done based on realm settings
UserRegistration --> validation logic server side
UserResource/AdminUI--> validation logic client side

This leads to the problem that if one merges all those logic into one place it has one of two possible consequences. 
1) Everywhere the same validation logic -> can break admin UI themes 
2) Add contexts to each validator e.g. UserProfileUpdate vs UserResourceUpdate vs UserRegistration vs ALL etc (not all known right now). This way validators can be specified for each context, which is especially useful for the legacy provider. 
The default context would be ALL and has not to be configured explicitly. 
This also implies that an attribute can have several instances of the same validator for different contexts. I don't expect to this kind of configuration in the wild but it's needed to fulfill the legacy compatibility. 

Validator config could look like this. 

```
 "validatiors": [{
                "name": { context:"UserProfileUpdate ", "length" : { "min" : 10, "max": 20 } }
                 }],
```

Reply:
Having the context of the update makes a lot of sense. Would be good to list the contexts we expect somewhere in this doc.

To add to the legacy use-case I can definitively think of other use-cases for being aware of the context.

Example:
* Admin registers account with email only - user is then asked to enter firstname/lastname on first login
* User registers themselves with email, firstname and lastname

For the legacy provider though what I had in mind was that would just be hard-coded as a legacy provider rather than using the JSON profile. It would still need to know the context though to be able to set the correct mandatory attributes as well as validation. Another way of doing that could be to have a pre-defined default JSON profile as well as a pre-defined legacy JSON profile. Current realms during migration would get the legacy profile, while new realms would get the new default profile. We should then probably have an easy option in the admin UI to switch to default, as well as reset a custom profile to the default.

Reply:
We agree to this approach, We will favor a pre-defined legacy profile if possible. Right now we think there might be edge cases which can't be overcome with the config approach (e.g. lagacy persistence logic) but we'll try to do so. 

As result it might turn out that we can only offer a "as close as it gets" config for the lagcy profile which can be used by people which want to keep the old behavoir. But this will may still have some changes in the details we have to communicate in the release updates.